File: crates/end2-wasm-client/frontend/src/types/message.ts
export type DecryptedMessage = {
    message_id: string,
    channel_id: string,
    author_id: string,
    plaintext: string,
    timestamp: Date,
}

export type InboundChatMessage = {
    message_id: string,
    author_id: string,
    device_id: string,
    channel_id: string,
    ciphertext: string,
    timestamp: string,
    is_pre_key: boolean,
}

export type EncryptedMessagePayload = {
    recipient_device_id: string,
    ciphertext: string,
    is_pre_key: boolean,
}

export type EncryptedMessage = {
    message_id: string,
    device_id: string,
    channel_id: string,
    payloads: EncryptedMessagePayload[],
}

export type MessageReceivedEvent = {
    message_id: string,
    channel_id: string,
    timestamp: string,
}

File: crates/end2-wasm-client/src/message.rs
use serde::{Deserialize, Serialize};
use time::OffsetDateTime;

use crate::types::{ChannelId, DeviceId, MessageId, UserId};

#[derive(Deserialize)]
pub struct InboundChatMessage {
    pub message_id: MessageId,
    pub device_id: DeviceId,
    pub channel_id: ChannelId,
    pub ciphertext: String, // b64encoded
    #[serde(with = "time::serde::rfc3339")]
    pub timestamp: OffsetDateTime,
    pub is_pre_key: bool,
}

#[derive(Clone, Deserialize, Serialize)]
pub struct DecryptedMessage {
    pub message_id: MessageId,
    pub channel_id: ChannelId,
    pub author_id: UserId,
    pub plaintext: String,
    #[serde(with = "time::serde::rfc3339")]
    pub timestamp: OffsetDateTime,
}

#[derive(Serialize)]
pub struct MessagePayload {
    pub recipient_device_id: DeviceId,
    pub ciphertext: String, // b64encoded
    pub is_pre_key: bool,
}
